libunwind/src/DwarfInstructions.hpp

```diff
@@ -403,9 +567,38 @@ int DwarfInstructions<A, R>::stepWithDwarf(A &addressSpace, pc_t pc,
       }
 #endif
 
+#if defined(__CHERI_PURE_CAPABILITY__) && defined(_LIBUNWIND_SANDBOX_OTYPES)
+      // If the sealer is not -1 (only the case when we're running with c18n),
+      // check if the return address has the executive mode bit set. If so, we
+      // should be calling into the c18n RTLD as this is a compartment boundary.
+      // We need to restore registers from the executive stack and ask rtld for
+      // it.
+      if (sealer != (uintcap_t)-1) {
+        // Iteratively unwind all the executive mode return addresses. This is
+        // necessary to support tail calls to trampolines.
+        uintcap_t csp = registers.getUnsealedECSP(sealer);
+	CompartmentInfo &CI = CompartmentInfo::sThisCompartmentInfo;
+        for (;;) {
+          if (isEndOfExecutiveStack(csp, CI)) {
```

Comment:
Is it possible that we are at the end of the Executive stack and yet still have frames to unwind? For example, we are in RTLD and have not made any compartment transitions, so the Executive stack would contain no frame.

To handle situations like this, I guess we can do something like
```
while (!isEndOfExecutiveStack(...) && isTrampoline(...)) {
  returnAddress = ...
  csp = ...
}
```

Reply:
Is that a situation we would ever end up in with libunwind? For us to begin unwinding, someone would have to have called `unw_getcontext` and `unw_init_local`, presumably in one of the compartments.